### ISSUE TITLE
Panicked at lib/src/network/http.rs:203 on unwrap(self.http())

### ISSUE THREAD

**Issue by BlackYoup** [sozu-proxy/sozu#127]
This is kinda related to #77 

I was testing some websocket requests and got that panic. I don't really know why it happens. What I did was just triggering multiple websocket upgrade requests with cURL: `curl -i -N -H "Connection: Upgrade" -H "Upgrade: websocket" -H "Sec-WebSocket-Key: Qr08/DkHrrqo/rheX0qLjA==" -H "Sec-WebSocket-Version: 13" lolcat.st` while debugging another bug. 

And after some requests and `^C` on cURL, it crashes (like: execute cURL, `^C` and do it again). It seems like random.

I'm not sure if that's websocket specific. I used that sample for my tests: https://gist.github.com/martinsik/2031681

```
thread 'main' panicked at '/home/arnaud/Dev/sozu/lib/src/network/http.rs:203 sozu_lib::network::http this should not have panicked:
tried to unwrap `None` in:
unwrap_msg!(self.http())', /home/arnaud/Dev/sozu/lib/src/util.rs:18
stack backtrace:
```

**Comment by Geal** [comment]
with 677e281 it should not crash anymore, but this is a case that should not happen: the proxy called `remove_backend` while it should have started to upgrade the connection. `http()` returns `None` because we're already at a different protocol. Do you have more logs for this?

**Comment by BlackYoup** [comment]
Here are more logs: https://gist.github.com/BlackYoup/a5692c7fac7203d7c115f7fbcf727676

Not sure how it happened, I noticed it 15 minutes later after it happened. Still during my websocket tests..

Will try to give more context if I reproduce

**Comment by BlackYoup** [comment]
Just got the same panick but with the TLS (or HTTPS) proxy:

```

**Comment by BlackYoup** [comment]
I just reproduced this bug with the HTTPS proxy. I refreshed a page while it was still loading. The webpage does a total of 64 requests in ~4.5s, all using SSL.

Here are some additional logs:
Sozu logs: https://gist.github.com/BlackYoup/c770665dcd7e525e5e009974955e1ad1 (full logs, I removed the certificate messages for obvious reasons)

coredump stacktrace:
```
Thread 1 (Thread 0x7fbfff3ff100 (LWP 3699)):
#0  0x00007fbffdff7ec6 in strlen () from /usr/x86_64-pc-linux-gnu/lib/libc.so.6
#1  0x000055b8f61a65f9 in std::ffi::c_str::CStr::from_ptr ()
#2  0x000055b8f618be11 in std::sys_common::gnu::libbacktrace::resolve_symname ()
#3  0x000055b8f6195026 in std::sys_common::backtrace::print ()
#4  0x000055b8f61a0e04 in std::panicking::default_hook::{{closure}} ()
#5  0x000055b8f61a0a5c in std::panicking::default_hook ()
#6  0x000055b8f61a12ce in std::panicking::rust_panic_with_hook ()
#7  0x000055b8f61a1193 in std::panicking::begin_panic ()
#8  0x000055b8f61a110d in std::panicking::begin_panic_fmt ()
#9  0x000055b8f6007c44 in sozu_lib::util::{{impl}}::unwrap_log<&mut sozu_lib::network::protocol::http::Http<openssl::ssl::SslStream<mio::net::tcp::TcpStream>>> (self=..., file=..., line=239, 
    module_path=..., expression=...) at /root/sozu/lib/src/util.rs:18
#10 0x000055b8f605de37 in sozu_lib::network::tls::{{impl}}::front_hup (self=0x7fbffaa63d18) at /root/sozu/lib/src/network/tls.rs:239
#11 0x000055b8f6096586 in sozu_lib::network::session::Session<sozu_lib::network::tls::ServerConfiguration, sozu_lib::network::tls::TlsClient>::ready<sozu_lib::network::tls::ServerConfiguration,sozu_lib::network::tls::TlsClient> (self=0x7ffebd9f3250, poll=0x7ffebd9f7e00, token=..., events=...) at /root/sozu/lib/src/network/session.rs:528
#12 0x000055b8f6086795 in sozu_lib::network::proxy::Server::run (self=0x7ffebd9f7e00) at /root/sozu/lib/src/network/proxy.rs:151
#13 0x000055b8f5affd44 in sozu::worker::begin_worker_process (fd=5, id=..., channel_buffer_size=10000) at /root/sozu/bin/src/worker.rs:96
#14 0x000055b8f5b07c15 in sozu::main () at /root/sozu/bin/src/main.rs:76
#15 0x000055b8f61ae8d1 in __rust_maybe_catch_panic ()
#16 0x000055b8f61a0fc8 in std::panicking::try ()
#17 0x000055b8f61a90b4 in std::rt::lang_start ()
#18 0x000055b8f5b13ad3 in main ()
```

I can reproduce the bug on demand and I'm using last master.

**Comment by Geal** [comment]
This should be fixed with 8afca1e

**Comment by BlackYoup** [comment]
I can't reproduce after this commit 👍
